src/common/Promise.ts

```diff
@@ -104,254 +82,30 @@ export class PromiseResultEventSource<T>  {
-                    if (!continuationPromise) {
-                        throw new Error("Continuation callback did not return promise");
-                    }
-                    continuationPromise.on((continuationResult: TContinuationResult) => {
-                        continuationDeferral.resolve(continuationResult);
-                    }, (e: string) => {
-                        continuationDeferral.reject(e);
-                    });
-                } catch (e) {
-                    continuationDeferral.reject(`'Error handler for error ${error} threw error ${e}'`);
-                }
-            },
-        );
-
-        return continuationDeferral.promise();
-    }
-
-    public onSuccessContinueWithPromise = <TContinuationResult>(
-        continuationCallback: (result: T) => Promise<TContinuationResult>): Promise<TContinuationResult> => {
-
-        if (!continuationCallback) {
-            throw new ArgumentNullError("continuationCallback");
-        }
-
-        const continuationDeferral = new Deferred<TContinuationResult>();
-
-        this.privSink.on(
-            (r: T) => {
-                try {
-                    const continuationPromise: Promise<TContinuationResult> = continuationCallback(r);
-                    if (!continuationPromise) {
-                        throw new Error("Continuation callback did not return promise");
-                    }
-                    continuationPromise.on((continuationResult: TContinuationResult) => {
-                        continuationDeferral.resolve(continuationResult);
-                    }, (e: string) => {
-                        continuationDeferral.reject(e);
-                    });
-                } catch (e) {
-                    continuationDeferral.reject(e);
-                }
-            },
-            (error: string) => {
-                continuationDeferral.reject(error);
-            },
-        );
-
-        return continuationDeferral.promise();
-    }
-
-    public on = (
-        successCallback: (result: T) => void,
-        errorCallback: (error: string) => void): Promise<T> => {
-        if (!successCallback) {
-            throw new ArgumentNullError("successCallback");
-        }
-
-        if (!errorCallback) {
-            throw new ArgumentNullError("errorCallback");
-        }
-
-        this.privSink.on(successCallback, errorCallback);
-        return this;
-    }
-
-    public finally = (callback: () => void): Promise<T> => {
-        if (!callback) {
-            throw new ArgumentNullError("callback");
-        }
-
-        const callbackWrapper = (_: any) => {
-            callback();
-        };
-
-        return this.on(callbackWrapper, callbackWrapper);
-    }
-}
-
 // tslint:disable-next-line:max-classes-per-file
 export class Deferred<T> implements IDeferred<T> {
     private privPromise: Promise<T>;
-    private privSink: Sink<T>;
+    private privResolve: (value?: T | PromiseLike<T>) => void;
```

Comment:
From what I'm reading, the "Like" types are for broad implementation compatibility. But why the use of PromiseLike here but not below on the 'resolve'?

    public resolve = (result: T | Promise<T>): Deferred<T> => {